### PR TITLE
Dataset Checksum

### PR DESCRIPTION
**Pull request by VJalili**
This PR calculates an MD5-based checksum for every uploaded dataset, and stores it as an attribute of a dataset in `dataset` table. 

### Why are we doing this? 
We are planning to enable data mounting, i.e., plug-and-play any storage/folder such as Amazon S3 bucket and Microsoft Azure blob. One of the basic requirements of this goal is to check and maintain data integrity. In other words, make sure a previously called `/a/path/someFile.bed` is the very same file the next time you access it (and probably, allow user to relocated a renamed and/or moved file). A common approach for data integrity assessment is using __checksum__--a (cryptographic) hash key generated based on a given file content. 

### What's next? 
Calculating checksum could be an expensive process as the entire file should be streamed in memory, which is a bottleneck in particular with large files. To avoid such challenges, in this PR we are streaming a file in memory in smaller chunks. However, a reasonable optimization would be parallelizing the upload and checksum calculation. 

Additionally, we also need to remotely compute checksum for remote files (e.g., hosted on S3). You may track a discussion on this topic [here](https://github.com/gvlproject/cloudbridge/issues/76) .

### Which datasets will be "assigned" a checksum ?
- Automatic: 
-- all datasets uploaded to or downloaded from a cloud-based storage (e.g., Amazon S3).
- Optional (user preference): 
-- all the dataset persisted on a local storage. 
- Excluded: 
-- all temporary datasets (e.g., workflow/tool execution related tmp files).

### When a checksum is computed by Galaxy ? 
- For data hosted on cloud: we first try to fetch a hash computed by the provider (e.g., boto ETag), if available we use it (most probably it is always available), if not Galaxy computes the checksum. 
- For data hosted on local storage: Galaxy computes the checksum, if set to (see above).

### If a checksum should be computed, which hash function is used ? 
- For data hosted on cloud: the hash function which is supported/required by the provider.
- For data hosted on a local storage: user-preference.

### CODE REVIEW

File: lib/galaxy/tools/actions/upload_common.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 import socket
 import subprocess
 import tempfile
+import hashlib
```
Comment (nsoranzo) [inline lib/galaxy/tools/actions/upload_common.py]: This import is not in alphabetical order, it should go first.

**Review by nsoranzo — COMMENTED** [review]
Is there a plan to also calculate the checksum for datasets created as tool outputs?

File: lib/galaxy/tools/actions/upload_common.py (hunk 1 above)
Comment (VJalili) [reply lib/galaxy/tools/actions/upload_common.py]: Thanks. Changed.

**Comment by martenson** [comment]
What about the data that is already in the DB, are we going to need checksums for all?

**Comment by VJalili** [comment]
That is a good point, and actually I'm working on it now.

An open challenge with that is: depending on the size of datasets, it might take a considerable amount of time for Galaxy to load for the first time after updated with this code. So, I'm not sure what would be the best solution. 

That is why I created this PR creating checksum for new datasets only. 

Alternatively, we could have a background process creating checksum for all the previously uploaded dataset.

**Comment by martenson** [comment]
@VJalili note that we may be talking about petabytes of data we need checksums for.

**Comment by VJalili** [comment]
@martenson please see my updated message.

**Comment by martenson** [comment]
@VJalili Best approach here might be to have an independent script do this outside of Galaxy. You probably want to run this as close to the data as possible to limit the network load.

Also note @nsoranzo's comment - you probably want _all_ datasets to have checksum, not just uploaded ones.

**Comment by VJalili** [comment]
@nsoranzo Yes, all the datasets will eventually have a checksum.

@martenson a script is a good idea. Generally I prefer the transition to be automated so that any user/admin would not need to worry of such an update. Therefore, even if we write a script, it should run automatically. But given the potentially expensive process, I might drop the preference this time :)

**Comment by martenson** [comment]
@VJalili A defensive programming approach might be beneficial here - assuming you are not guaranteed the checksum exists. There are many fairly large Galaxies with years of history (and dataset production).

**Comment by jmchilton** [comment]
Very cool start - thanks so much! I have applications I'd like to use this for - I don't think I'd ever want to use md5sum as the hashing function though.

- Can you add columns for other checksums and suffix this one with md5? (e.g. checksum_md5, checksum_sha1, checksum_sha256 or perhaps use a separate table with columns for hash value and hash type).
- Can you add an option to Galaxy ``default_collect_hashes`` that is a list of hash functions to compute and default this to the empty list. We should verify this works and is useful with a particular application before applying it in a blanket fashion. Galaxy is already a very slow, cumbersome way to run jobs - this will make that a lot worse. It doesn't mean this is a bad idea - I just think we need to take a very light weight touch with this and add the complexity to jobs and applications that need it - not every job and every application.
- Can you include an application - something easy might be to include hashes with uploads to the API and add some API tests that allow a valid hash and a test that disallows an invalid hash. If a hash is included with an upload but isn't in the default_collect_hashes list - you can have Galaxy compute it on the fly in this fashion and populate the database entry. A cooler application would be sketch out Pulsar communication integrity checks or a rewrite of Pulsar's caching to use this. Sketching out hashing threads would be cool also - like job handlers or workflow schedulers - these could be threads that just hash data - if you had these hashing could happen in the background as needed for applications like Pulsar. You could then have a new job state also - "hashing_inputs" that would wait for the requested hashes to be ready before scheduling a job that uses the hash.

**Comment by VJalili** [comment]
Thanks @jmchilton I'll get to your points one-by-one as I develop. First point first: 

> Can you add columns for other checksums and suffix this one with md5? (e.g. checksum_md5, checksum_sha1, checksum_sha256 or perhaps use a separate table with columns for hash value and hash type).

I have been thinking on this for a while. As I said, the main motivation of checksum is to allow **mounting** data from cloud-based storage providers. However, these providers could potentially use different hashing functions, in particular when parallel uploading large files in chunks. Therefore, it would be difficult (if not impossible) to come with one hash function that matches the hash generated by all these providers. Therefore, using multiple hash functions sound quite promising. But the drawback of multiple hash functions is the escalated computational complexity. We're already having challenges with just one hash function, and computing multiple (e.g., md5, SHA-1, SHA256, SHA512) will be even more challenging. So, I'm wonder what would be a best approach balancing a compute time vs. portability tradeoff.

**Comment by dannon** [comment]
I'd agree we need either a separate table with algorithm + value, or to encode the algorithm along with it in the checksum column you have, so something like `sha1$da39a3ee5e6b4b0d3255bfef95601890afd80709`.

A benefit to having a separate table, is that you could store multiple potential checksums with different algorithms for a single dataset.

**Comment by VJalili** [comment]
@dannon separate table would be fine, but what do you think of the cost of computing multiple hash  keys ?

Second, what do you think of the following schema for a table: 

ID | dataset ID | md5 | SHA-1 | SHA256 | SHA512 

one column per each function vs the following schema: 

ID | dataset ID | hash function | hash value

**Comment by jmchilton** [comment]
> But the drawback of multiple hash functions is the escalated computational complexity. We're already having challenges with just one hash function, and computing multiple (e.g., md5, SHA-1, SHA256, SHA512) will be even more challenging.

I think by moving from a model where we compute the hash upfront to one where we can request the hash we need as we need it we avoid these tradeoffs to some degree. We are then never hashing something we don't request - so every hash that is generated was specifically requested and so is probably needed. If particular admins or particular applications want to pre-hash something - they could enable that as needed.

**Comment by dannon** [comment]
> ID | dataset ID | hash function | hash value

Is a whole lot more generalized, and what I had in mind.

Regarding the cost, I definitely would not suggest hashing all the different types on upload or whatever.  To my mind, the hash is primarily important when data crosses in/out of the Galaxy data boundary.  For data that never leaves, like workflow intermittent datasets?  We might (even probably) never need a hash at all, ever.  I'd say we should make hashes completely optional by implementation, with a later pass maybe adding configurable 'always compute a hash using choice_of_algorithm' setting or something later.

Edit: Pfft, comments crossed in the ether.  Yeah, I'm looking at this the same way @jmchilton is I think.

**Comment by mvdbeek** [comment]
> Regarding the cost, I definitely would not suggest hashing all the different types on upload or whatever.

For non-cryptographic hashes that should be mostly I/O right ? We could perhaps embed that with the actual job. Having hashes may come in handy for checking integrity of old datasets, de-duplication and other things we may not be thinking about now.

**Comment by dannon** [comment]
@mvdbeek If we picked a single ultra-low cost checksum, *maybe* it'd be worth doing always as a part of the job, on by default.  I think it's way safer to start from the position of making it completely optional, generated-on-request, and build up to that.  With lots of large files, it's non-trivial hardware time any way you look at it, for something that a private local internal-use-only install may not ever want, right?

Another thought is that it looks like for this to be very useful at least in the cloud context we're going to want to support multiple different algorithms.  We have to be able to handle the situation where the internal data does not have an available hash with `insert-fancy-new-algorithm`.  Since we're going to have to handle the case of hashes not being available, does it make sense to start with that as the default position?

**Comment by VJalili** [comment]
@dannon  @jmchilton  @mvdbeek Thanks for the comments. I tried to summarize the discussion as a plan to formally define when/where/how to compute checksum. Please check the updated first message. And comment if you prefer a change.

**Comment by martenson** [comment]
This is marked as WIP so I am pushing it to 18.05.

**Comment by martenson** [comment]
generalized in https://github.com/galaxyproject/galaxy/pull/7487

**Pull request merged**